### ISSUE TITLE
Cannot restore the original order using Shift + Click

### ISSUE THREAD

**Issue by yihui** [DataTables/DataTables#526]
Thanks for fixing #391 and I have discovered a related problem: if you click the table header holding Shift, you cannot really cancel sorting, i.e. the original order is not restored after the Up/Down arrows are gone. It is easy to reproduce the problem with any table, e.g. http://datatables.net/examples/basic_init/table_sorting.html You can see the original order in the HTML panel under the table, and you cannot go back to the original order using Shift + Click.

**Comment by DataTables** [comment]
The bug here is really that you shouldn't be able to shift click to remove sorting. There shouldn't be any option to remove sorting by clicking on the headers.

Probably not what you wanted to hear I guess, but that is what my intention was.

**Comment by DataTables** [comment]
Fixed in the above commit.

**Comment by yihui** [comment]
Fair enough, I successfully shot myself in the foot :sob: 

I fully respect your decision, but it is actually a _very_ nice feature to be able to cancel sorting and restore to the original order. I believe I'm not alone here: https://www.google.com/webhp?sourceid=chrome-instant&ion=1&espv=2&ie=UTF-8#q=datatables%20reset%20sorting It is too bad that we can never go back after we sort a column.

You also have a plug-in for this task: http://datatables.net/plug-ins/api/fnSortNeutral I will truly appreciate it if this can be a built-in feature.

**Comment by DataTables** [comment]
Yes, there is a plug-in for this, the old `fnSortNeutral` and the [new `order.neutral()`](https://github.com/DataTables/Plugins/blob/master/api/order.neutral().js) methods (which will be deployed to the DataTables site soon).

I don't currently have any plans to integrate this ability into the core.

**Comment by yihui** [comment]
Thanks for the pointer to the new method. I tried `order.neutral()` and it does not seem to work correctly. You can reproduce the problem by opening an arbitrary example, say, http://datatables.net/examples/basic_init/zero_configuration.html Register the plug-in, and run

```js
$('#example').DataTable().order.neutral().draw();
```

The order seems to be removed, but you see Jena Gaines is the third person in the table, which actually should be Ashton Cox (looking at the HTML source table).

Back to my request, I think it is 1) super useful to restore the order in the case of a single ordered column, and 2) confusing when we are able to remove the ordering in case of multiple columns but are not able to do so in case of a single column. I wonder what your concerns are.

Prior to this fix, I could at least tell [my package](http://rstudio.github.io/DT) users that it was possible to remove ordering using the server-side processing mode, because the original data is stored on the server (in R), and when the order is removed, I can send the original data back. Now you have disabled this feature, I will have to lose my last solution.

Even if we have got a plug-in working correctly, the question is what the best way is to trigger the neutral order: I think Shift + Click is a perfect place to call `order.neutral()`. It will be awkward to provide a separate button such as `Remove Sorting` in the UI that does this job.

**Comment by DataTables** [comment]
Thanks for pointing out the error with the `order.neutral()` plug-in. I've committed the fix](https://github.com/DataTables/Plugins/commit/d1ba8bde45e2ee).

As to why I don't currently thing this should be in the core, removing sorting from the table isn't something that I would find useful int he tables I have created myself, and it is a relatively unusual feature. Having said that, I don't have a problem with including it as such - if it could be done with a little less code that the plug-n uses (I'm not sure why it needs the inner sorting function...) then it is something I would consider.

I'll repoen and mark this as a feature.

**Comment by yihui** [comment]
Excellent! Tears rolling down my face again :joy:

**Comment by yihui** [comment]
Just a polite annual ping :smiley:

**Comment by DataTables** [comment]
Hi - thanks for the ping. This isn't something I feel I'm likely to include in the DataTables core at the moment. There is the plug-in method and I will look at adding a way to call external methods such as that for the next major version, but the ability to remove ordering completely from a table isn't a request that comes up very often, and I think could potentially be confusing to some.

**Comment by yihui** [comment]
I do not see how it could be confusing, but anyway, I think it will be nice to make this configurable, e.g. through an option `orderRestore` (`false` by default).

**Comment by Alex1990** [comment]
:+1:

**Comment by dhobi** [comment]
:+1:

**Comment by Stadly** [comment]
:+1:

**Comment by bwganblack** [comment]
![](https://assets-cdn.github.com/images/icons/emoji/unicode/1f44d.png)

**Comment by Gerard987789** [comment]
If I select data from MySQL it will be sorted on id. But in DataTables I am not showing the id column as it does not have any value for my clients. When data is arriving at the client it is ordered by id and than if the client is clicking on one of the headers they never will get the original order back again.
That is for my a reason to support the request for a 3 state; up-down-neutral.
